### DISCUSSION TITLE
No postcode after adding a second region

### DISCUSSION THREAD

**Discussion by vhmoura** [osm-search/Nominatim#2421] [Geocoding around the world]
Hi all,
I am currently running the street map locally where I initially added UK and Ireland maps. Everything works as expected and I can reverse geocode lat/longs with no issues.
I added a second regions, South Africa to do the same reverse geocoding. I followed the steps provided in my previous discussion from here: https://www.nominatim.org/release-docs/3.6.0/admin/Advanced-Installations/#importing-multiple-regions

I managed to import South Africa map and run the update. However, running a query like this: reverse.php?format=jsonv2&lat=-25.769930&lon=28.382810&addressdetails=1&extratags=1&zoom=17 

returns no post code in the address detail.

Copying and paste the very same address above to Nominatim website https://nominatim.openstreetmap.org/reverse.php?format=jsonv2&lat=-25.769930&lon=28.382810&addressdetails=1&extratags=1&zoom=17

return the correct post code in the address detail. 

Did I perhaps missed a step during import? 
I have also done compute postcodes but it doesn't make any difference.

Any help appreciated.
Cheers
Vini

**Comment by mtmail** [comment]
Can you check if there's an entry for the postcode in the `location_postcode` table?


```sql
SELECT * FROM location_postcode WHERE country_code='za' AND postcode='0054';
place_id        | 282817465
parent_place_id | 280470792
rank_search     | 21
rank_address    | 11
indexed_status  | 0
indexed_date    | 2021-08-03 17:25:43.131053
country_code    | za
postcode        | 0054
geometry        | 0101000020E6100000C6DF440D4A5F3C40EA57B364EBC239C0
```

```sql
SELECT count(*) FROM location_postcode WHERE country_code='za';
773
```

The `cd build; ./utils/update.php --calculate-postcodes` command updates this table (https://www.nominatim.org/release-docs/3.6.0/develop/Postcodes/), you said you already ran it.

It might be necessary to reindex the places, too. Try one place first, for example "City of Tshwane Metropolitan Municipality" https://www.openstreetmap.org/relation/594484

1. `SELECT place_force_update((SELECT place_id FROM placex WHERE osm_type='R' and osm_id='594484'));`
2. `cd build; ./utils/update.php --index`

**Comment by lonvia** [comment]
That's tricky. The artificial postcodes need to be updated before indexing. So when adding a file, the order of execution should be:

```
nominatim add-data --file <FILE>
nominatim refresh --postcodes
nominatim index
```

We should update the documentation accordingly.

**Reply by vhmoura** [reply]
Hi Lonvia, 
I am still facing issues with postcode. I have done all the steps you mentioned here but with no results. 
I have recently imported a new map (NZ) and using the exactly same search for both nominatim and my own, returns different results: Nominatim returns postcode, but not my solution. 
Can you point where in the php files the postcode lookup happens?

One thing that I noticed though (which might be the problem),  the place_id and parent_id returned seems different than what I have in the lookup database, please see below:

`Places:  103597666 => 'osm_type' => 'N'
                      'osm_id' => 5683477800
                      'class' => 'place'
                      'type' => 'house'
                      'admin_level' => 15
                      'rank_search' => 30
                      'rank_address' => 30
                      'place_id' => 103597666
                      'parent_place_id' => 103921629
                      'housenumber' => -1
                      'country_code' => 'nz'
                      'langaddress' => '257, Manuherikia Road, Alexandra, Vincent Community, Central Otago District, Otago, New Zealand'
                      'placename' => 
                      'ref' => 
                      'lon' => '169.4222771'
                      'lat' => '-45.2315454'
                      'importance' => 0
                      'addressimportance' => '0.00034045053971839395'
                      'extra_place' => 
                      'addresstype' => 'place'`


<img width="1010" alt="image" src="https://user-images.githubusercontent.com/20772905/184614757-f350d6f5-4379-44de-9b19-7bededd51eba.png">


And you can see here the different when making both calls:

<img width="849" alt="image" src="https://user-images.githubusercontent.com/20772905/184622205-25ee2a75-3ff7-4720-977c-584173fdb08e.png">


<img width="815" alt="image" src="https://user-images.githubusercontent.com/20772905/184622019-1e7217f0-d9b3-4ec1-a1bd-cab750723883.png">


Can you please help?

**Reply by mtmail** [reply]
`place_id` is different for each import and will differ between servers, that's expected https://nominatim.org/release-docs/develop/api/Output/#place_id-is-not-a-persistent-id

Which version of Nominatim are you using?

Can you see an entry in the `location_postcode` table for the postcode?

```sql
select *,st_astext(geometry) from location_postcode where country_code='nz' and postcode='9393';
```

**Reply by vhmoura** [reply]
Hi,

I would expect that importing a new country would create ids that are related across the tables. What I am inferring is that the place id is different in the postcode table and place table, which should be the same shouldn't (or is it not related)? anyhow, running the query does return a record.

**Reply by mtmail** [reply]
Each place has a `place_id`, a postcode is its own place so the ids in the `location_postcode` table are different from anything else in the `placex` table. I wouldn't worry about the ids, they're just an internal data structure.

When Nominatim does a reverse search it looks at the exact search coordinates first, then increases the search radius until it finds a place (e.g. a fountain, a restaurant, a house number, a street, a city). Then for that place it looks up it's address as each street belongs to one neighborhood, each neighborhood to one city, each city to one state etc. Postcodes are separate though. That's another lookup SQL query and a postcodes doesn't belong to streets or cities or states, they're outside the address hierarchy.

What is the coordinate for the 9393 postcode in the `location_postcode` table? In my database I have `POINT(169.41131921024285 -45.2381739)` (map: https://nominatim.openstreetmap.org/ui/reverse.html?lat=-45.2381739&lon=169.41131921024285&zoom=18). Postcodes in Nominatim are "only" center points. The center points are calculated from places in the `placex` table (the `postcode` column among other fields). It's possible that your center point is different, further away and then wouldn't match.

Does https://nominatim.openstreetmap.org/ui/search.html?q=9393%2C+nz return a postcode result on your server?

**Reply by vhmoura** [reply]
Hi,

Thanks for explaining about the ids.

Please see below the image when requesting the location_postcode - same as you

<img width="1265" alt="image" src="https://user-images.githubusercontent.com/20772905/184646107-a36fe635-cb1a-4b81-b3ed-51bd2bd5ac6f.png">

and here running the search:

<img width="801" alt="image" src="https://user-images.githubusercontent.com/20772905/184646737-55a8e1b8-352f-4bbe-a921-41a6c4d30c2d.png">


is it possible increase the center point? if so, where should I do it?

**Comment by lonvia** [comment]
I tried the following now using Nominatim 4.1.0:

1. Import Monaco into a fresh database: `nominatim import --osm-file monaco-latest.osm.pbf`
2. Add Liechtenstein to my database:
```
nominatim add-data --file liechtenstein-latest.osm.pbf
nominatim refresh --postcodes
nominatim index
```

After this is done, I have Liechtenstein data with postcodes:

```
http://localhost/reverse.html?lat=47.11715&lon=9.54380&zoom=16
{
  "place_id": 258347,
  "licence": "Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright",
  "osm_type": "way",
  "osm_id": 60827687,
  "lat": "47.1171808",
  "lon": "9.543818",
  "place_rank": 26,
  "category": "highway",
  "type": "residential",
  "importance": 0.09999999999999998,
  "addresstype": "road",
  "name": "Rossbodastrasse",
  "display_name": "Rossbodastrasse, Steinort, Rotenboden, Triesenberg, Oberland, 9497, Liechtenstein",
  "address": {
    "road": "Rossbodastrasse",
    "hamlet": "Steinort",
    "village": "Rotenboden",
    "county": "Oberland",
    "postcode": "9497",
    "country": "Liechtenstein",
    "country_code": "li"
  },
  "boundingbox": [
    "47.1139377",
    "47.1173403",
    "9.543645",
    "9.5468163"
  ]
}
```

**Reply by vhmoura** [reply]
I am using nominatim 3.6

what could be causing the postcode not to appear in the result? What part of the php does that so I can try to fix myself?

Many thanks

**Reply by mtmail** [reply]
https://github.com/osm-search/Nominatim/tree/v3.6.0/lib

If I recall 3.x versions correct `ReverseGeocode.php` calls `lookup()` function in `PlaceLookup.php`. That's where you find a section "postcode table". The `Debug::printVar(` statements are printed in a HTML when you add `&debug=1` to a URL, e.g. https://nominatim.openstreetmap.org/search.php?q=9393%2C+nz&format=jsonv2&debug=1

Check your server's output of https://nominatim.openstreetmap.org/reverse.php?format=jsonv2&lat=-45.2318152&lon=169.422485&addressdetails=1&extratags=1&zoom=17&debug=1 and you will see the relevant SQL queries.

**Reply by lonvia** [reply]
The postcodes are added on import, so the PHP for queries is completely irrelevant here.

Nominatim 3.6 has a completely different set of commands for import, so your claim "I have done all the steps you mentioned" was obviously wrong. If I remember correctly, then 3.6 doesn't have an update function for postcode, so there is absolutely nothing you can do except using a more recent version of Nominatim.

**Reply by mtmail** [reply]
Hmm, true, `nominatim refresh --postcodes` didn't exist in Nominatim 3.6

**Reply by vhmoura** [reply]
Fair enough. Would you guys recommend updating from 3.6->3.7>4>4.1 or just do a new install on 4.1?

Thanks for your help.

**Reply by mtmail** [reply]
Starting Nominatim 3.7 the database can be migrated with the `nominatim admin --migrate` command. https://nominatim.org/release-docs/latest/admin/Migration/

I'd still do a new install of 4.1. Most install commands changed, the directory paths changed (new project directory) and the configuration file has a new format compared to 3.6.
